test/switch_test.dart

```diff
@@ -113,6 +113,25 @@ void main() {
           expect(firstCanceled, true);
         });
 
+        if (innerType != 'broadcast') {
+          test('waits for cancel before listening to subsequent stream',
+              () async {
+            var cancelWork = Completer<void>();
+            first.onCancel = () => cancelWork.future;
+            outer.add(first.stream);
+            await Future(() {});
+
+            var cancelDone = false;
+            second.onListen = expectAsync0(() {
+              expect(cancelDone, true);
+            });
+            outer.add(second.stream);
```

Comment:
Try adding a third stream here as well, and ensure that the behavior is still desirable (and predictable, so document what happens).